From 8eb423d2792f7634d838a0e3b57eebcffefe12df Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?M=C3=A9ven=20Car?= <meven@kde.org>
Date: Mon, 6 Feb 2023 02:32:32 +0100
Subject: [PATCH] Mobi extractor: only extract what is asked

BUG: 465006
---
 src/extractors/mobiextractor.cpp | 56 +++++++++++++++++---------------
 1 file changed, 29 insertions(+), 27 deletions(-)

diff --git a/src/extractors/mobiextractor.cpp b/src/extractors/mobiextractor.cpp
index d66967a..0aa9288 100644
--- a/src/extractors/mobiextractor.cpp
+++ b/src/extractors/mobiextractor.cpp
@@ -58,36 +58,38 @@ void MobiExtractor::extract(ExtractionResult* result)
 
     result->addType(Type::Document);
 
-    QMapIterator<Mobipocket::Document::MetaKey, QString> it(doc.metadata());
-    while (it.hasNext()) {
-        it.next();
-        switch (it.key()) {
-        case Mobipocket::Document::Title:
-            result->add(Property::Title, it.value());
-            break;
-        case Mobipocket::Document::Author: {
-            result->add(Property::Author, it.value());
-            break;
-        }
-        case Mobipocket::Document::Description: {
-            QTextDocument document;
-            document.setHtml(it.value());
-
-            QString plain = document.toPlainText();
-            if (!plain.isEmpty())
-                result->add(Property::Description, it.value());
-            break;
-        }
-        case Mobipocket::Document::Subject:
-            result->add(Property::Subject, it.value());
-            break;
-        case Mobipocket::Document::Copyright:
-            result->add(Property::Copyright, it.value());
-            break;
+    if (result->inputFlags() & ExtractionResult::ExtractMetaData) {
+        QMapIterator<Mobipocket::Document::MetaKey, QString> it(doc.metadata());
+        while (it.hasNext()) {
+            it.next();
+            switch (it.key()) {
+            case Mobipocket::Document::Title:
+                result->add(Property::Title, it.value());
+                break;
+            case Mobipocket::Document::Author: {
+                result->add(Property::Author, it.value());
+                break;
+            }
+            case Mobipocket::Document::Description: {
+                QTextDocument document;
+                document.setHtml(it.value());
+
+                QString plain = document.toPlainText();
+                if (!plain.isEmpty())
+                    result->add(Property::Description, it.value());
+                break;
+            }
+            case Mobipocket::Document::Subject:
+                result->add(Property::Subject, it.value());
+                break;
+            case Mobipocket::Document::Copyright:
+                result->add(Property::Copyright, it.value());
+                break;
+            }
         }
     }
 
-    if (!doc.hasDRM()) {
+    if ((result->inputFlags() & ExtractionResult::Flag::ExtractPlainText) && !doc.hasDRM()) {
         QString html = doc.text();
 
         QTextDocument document;
-- 
GitLab

